### PR TITLE
*: add region option for random pick region

### PR DESCRIPTION
**Pull request by nolouch**
same as the title.

### CODE REVIEW

File: server/cache.go [hunk 1]
```diff
@@ -279,10 +279,10 @@ func (c *clusterInfo) getRegions() []*core.RegionInfo {
 	return c.Regions.GetRegions()
 }
 
-func (c *clusterInfo) randomRegion() *core.RegionInfo {
+func (c *clusterInfo) randomRegion(ops ...core.RegionOption) *core.RegionInfo {
```
Comment (Connor1996) [inline server/cache.go]: use `opts` is less confusing。

**Comment by disksing** [comment]
Can we abstract 'has leader on store x' and 'has follower on store x' as options?

File: server/core/region.go [hunk 2]
```diff
@@ -26,6 +26,19 @@ import (
...
+// RegionOption used to select region
+type RegionOption func(region *RegionInfo) bool
+
+// HealthRegion checks if the region is healthy
+func HealthRegion() RegionOption {
+	return func(region *RegionInfo) bool {
+		if len(region.DownPeers) > 0 || len(region.PendingPeers) > 0 {
+			return false
+		}
+		return true
```
Comment (disksing) [inline server/core/region.go]:
```
return len(downPeers) == 0 && len(pendingPeers) == 0
```

File: server/cache.go [hunk 3]
```diff
@@ -324,17 +324,17 @@ func (c *clusterInfo) getStoreLeaderCount(storeID uint64) int {
 }
 
 // RandLeaderRegion returns a random region that has leader on the store.
-func (c *clusterInfo) RandLeaderRegion(storeID uint64) *core.RegionInfo {
+func (c *clusterInfo) RandLeaderRegion(storeID uint64, ops ...core.RegionOption) *core.RegionInfo {
```
Comment (Connor1996) [inline server/cache.go]: s/ops/opts

File: server/cache.go [hunk 4]
```diff
@@ -324,17 +324,17 @@ func (c *clusterInfo) getStoreLeaderCount(storeID uint64) int {
...
 	c.RLock()
 	defer c.RUnlock()
-	return c.BasicCluster.RandLeaderRegion(storeID)
+	return c.BasicCluster.RandLeaderRegion(storeID, ops...)
 }
 
 // RandFollowerRegion returns a random region that has a follower on the store.
-func (c *clusterInfo) RandFollowerRegion(storeID uint64) *core.RegionInfo {
+func (c *clusterInfo) RandFollowerRegion(storeID uint64, ops ...core.RegionOption) *core.RegionInfo {
```
Comment (Connor1996) [inline server/cache.go]: ditto

File: server/namespace_cluster.go [hunk 5]
```diff
@@ -60,9 +60,9 @@ func (c *namespaceCluster) checkRegion(region *core.RegionInfo) bool {
 const randRegionMaxRetry = 10
 
 // RandFollowerRegion returns a random region that has a follower on the store.
-func (c *namespaceCluster) RandFollowerRegion(storeID uint64) *core.RegionInfo {
+func (c *namespaceCluster) RandFollowerRegion(storeID uint64, ops ...core.RegionOption) *core.RegionInfo {
```
Comment (Connor1996) [inline server/namespace_cluster.go]: ditto.

File: server/namespace_cluster.go [hunk 6]
```diff
@@ -74,9 +74,9 @@ func (c *namespaceCluster) RandFollowerRegion(storeID uint64) *core.RegionInfo {
 }
 
 // RandLeaderRegion returns a random region that has leader on the store.
-func (c *namespaceCluster) RandLeaderRegion(storeID uint64) *core.RegionInfo {
+func (c *namespaceCluster) RandLeaderRegion(storeID uint64, ops ...core.RegionOption) *core.RegionInfo {
```
Comment (Connor1996) [inline server/namespace_cluster.go]: ditto

File: server/schedule/basic_cluster.go [hunk 7]
```diff
@@ -146,13 +146,13 @@ func (bc *BasicCluster) UnblockStore(storeID uint64) {
 }
 
 // RandFollowerRegion returns a random region that has a follower on the store.
-func (bc *BasicCluster) RandFollowerRegion(storeID uint64) *core.RegionInfo {
-	return bc.Regions.RandFollowerRegion(storeID)
+func (bc *BasicCluster) RandFollowerRegion(storeID uint64, ops ...core.RegionOption) *core.RegionInfo {
```
Comment (Connor1996) [inline server/schedule/basic_cluster.go]: ditto

File: server/schedule/basic_cluster.go [hunk 8]
```diff
@@ -146,13 +146,13 @@ func (bc *BasicCluster) UnblockStore(storeID uint64) {
...
+	return bc.Regions.RandFollowerRegion(storeID, ops...)
 }
 
 // RandLeaderRegion returns a random region that has leader on the store.
-func (bc *BasicCluster) RandLeaderRegion(storeID uint64) *core.RegionInfo {
-	return bc.Regions.RandLeaderRegion(storeID)
+func (bc *BasicCluster) RandLeaderRegion(storeID uint64, ops ...core.RegionOption) *core.RegionInfo {
```
Comment (Connor1996) [inline server/schedule/basic_cluster.go]: ditto

**Pull request merged**